test/test_filter_map.cpp

```diff
@@ -0,0 +1,164 @@
+
+// Copyright (c) 2024 Tristan Brindle (tcbrindle at gmail dot com)
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#include "catch.hpp"
+
+#include <array>
+#include <utility>
+
+#include "test_utils.hpp"
+
+namespace {
+
+constexpr auto is_even_opt = [](int i) { return i % 2 == 0 ? std::optional{i} : std::nullopt; };
+
+struct Pair {
+    int i;
+    bool ok;
+
+    [[nodiscard]] constexpr auto map_if_ok() const { return ok ? std::optional{*this} : std::nullopt; }
+
+    constexpr bool operator==(const Pair&) const = default;
+};
+
+using filter_fn = decltype(flux::filter_map);
+
+// int is not a sequence
+static_assert(not std::invocable<filter_fn, int, decltype(is_even_opt)>);
+// int is not a function
+static_assert(not std::invocable<filter_fn, int(&)[10], int>);
+// "func" does not return optional_like
+static_assert(not std::invocable<filter_fn, int(&)[10], decltype([](int) {})>);
+// Incompatible predicate
+static_assert(not std::invocable<filter_fn, int(&)[10], decltype([](int*) { return true; })>);
+
+constexpr bool test_filter()
+{
+    // Basic filtering
+    {
+        int arr[] = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
+        auto filtered = flux::filter_map(flux::ref(arr), is_even_opt);
+        using F = decltype(filtered);
+        static_assert(flux::sequence<F>);
+        static_assert(flux::bidirectional_sequence<F>);
+        static_assert(flux::bounded_sequence<F>);
+        static_assert(not flux::ordered_cursor<F>);
+        static_assert(not flux::sized_sequence<F>);
+
+        static_assert(flux::sequence<F const>);
+        static_assert(flux::bidirectional_sequence<F const>);
+        static_assert(flux::bounded_sequence<F const>);
+        static_assert(not flux::ordered_cursor<F const>);
+        static_assert(not flux::sized_sequence<F const>);
+
+        STATIC_CHECK(check_equal(filtered, {0, 2, 4, 6, 8}));
+        STATIC_CHECK(check_equal(std::as_const(filtered), {0, 2, 4, 6, 8}));
+    }
+
+    // A predicate that always returns true returns what it was given
+    {
+        int arr[] = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
+        auto filtered = flux::filter_map(flux::ref(arr), [](auto&& i) { return std::optional{i}; });
+
+        if (!check_equal(arr, filtered)) {
+            return false;
+        }
```

Comment:
Please use the `STATIC_CHECK` macro rather than a manual `if` statement as it gives better diagnostics if there's a constexpr test failure (here and in other places in this file)

```suggestion
        STATIC_CHECK(check_equal(arr, filtered));
```

(I guess you copied this from `test_filter.cpp`, which was mostly written before the macro existed and should probably be updated to use it)